File: 3336-water-bottles-ii/water-bottles-ii.go
func maxBottlesDrunk(numBottles int, numExchange int) int {
	empty := numBottles
	drank := numBottles
	numBottles = 0
    
	for empty >= numExchange {
		// exchange bottles on by one while incrementing numExchange
		for empty >= numExchange {
			numBottles++
			empty -= numExchange
			numExchange++
		}

		// and drinking them again
		drank += numBottles
		empty += numBottles
		numBottles = 0
	}

	return drank
}

File: 3336-water-bottles-ii/water-bottles-ii.py
class Solution:
    def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
        res = numBottles
        emptyBottles = numBottles
        fullBottles = 0

        while numExchange <= emptyBottles:
            while emptyBottles >= numExchange:
                fullBottles += 1
                emptyBottles -= numExchange
                numExchange += 1

            res += fullBottles
            emptyBottles += fullBottles
            fullBottles = 0

        return res
